File: application/app/domain/model/Route.scala
package domain.model

import common.validation.Validatable
import common.validation.constraints.TraversableSize
import domain.exceptions.GenericException
import domain.model.RouteType.RouteType
import domain.validation.constraints.ExpressionByRouteType
import javax.validation.constraints.{NotBlank, NotNull, Size}

import scala.annotation.meta.field

@ExpressionByRouteType
case class Route(@(NotNull @field) routeType: RouteType,
                 @(Size @field)(min = 4, max = 255) @(NotBlank @field) name: String,
                 @(TraversableSize @field)(max = 10) backends: Set[Backend[BackendSettings]],
                 expression: Option[Expression[Boolean]]
                ) extends Validatable {

  override def hashCode(): Int = {
    val prime = 31
    var result = 1
    result = prime * result + routeType.hashCode()
    result = prime * result + (if (name == null) 0 else name.hashCode)
    result
  }

  override def equals(that: Any): Boolean = {
    that match {
      case that: Route => {
        that.canEqual(this) &&
        this.routeType == that.routeType &&
        this.name == that.name
      }
      case _ => false
    }
  }

  def appliesTo(json: String): Boolean = {
    if (routeType != RouteType.CUSTOM) {
      throw GenericException(s"expression can be evaluated only for custom routes")
    }
    expression match {
      case Some(e) => e.evaluate(json)
      case None => throw GenericException("Custom route does not have an expression")
    }
  }

  def canEqual(other: Any): Boolean = other.isInstanceOf[Route]
}

object Route {
  private val debugBackend: Set[Backend[BackendSettings]] = Set(Backend.debug())

  def connect(): Route = Route(RouteType.CONNECT, "Connect", debugBackend)
  def connect(backends: Set[Backend[BackendSettings]]): Route = Route(RouteType.CONNECT, "Connect", backends)

  def disconnect(): Route = Route(RouteType.DISCONNECT, "Disconnect", debugBackend)
  def disconnect(backends: Set[Backend[BackendSettings]]): Route = Route(RouteType.DISCONNECT, "Disconnect", backends)


  def default(): Route = Route(RouteType.DEFAULT, "Default route", debugBackend)
  def default(backends: Set[Backend[BackendSettings]]): Route = Route(RouteType.DEFAULT, "Default route", backends)

  def apply(routeType: RouteType, name: String): Route = new Route(routeType, name, Set.empty, Option.empty)
  def apply(routeType: RouteType, name: String, backends: Set[Backend[BackendSettings]]): Route = {
    new Route(routeType, name, backends, Option.empty)
  }
  def apply(routeType: RouteType, name: String,
            backends: Set[Backend[BackendSettings]],
            expression: Option[Expression[Boolean]]): Route = {
    new Route(routeType, name, backends, expression)
  }
}

File: ws-gateway-domain/src/main/java/com/cosmin/wsgateway/domain/Route.java
package com.cosmin.wsgateway.domain;

import com.cosmin.wsgateway.domain.exceptions.InvalidRouteException;
import com.cosmin.wsgateway.domain.validation.constraints.ExpressionByRouteType;
import java.util.Collections;
import java.util.Optional;
import java.util.Set;
import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Value;
import lombok.With;

@Value
@Builder
@ExpressionByRouteType
public class Route {
    public enum Type {
        CONNECT,
        DISCONNECT,
        DEFAULT,
        CUSTOM
    }

    @NotNull
    @Size(min = 4, max = 255)
    private final String name;

    @NotNull
    private final Type type;

    @Size(max = 10)
    @Valid
    @NotNull
    @EqualsAndHashCode.Exclude
    @With
    private final Set<Backend<? extends BackendSettings>> backends;

    @EqualsAndHashCode.Exclude
    @With
    @Valid
    private final Optional<Expression<Boolean>> expression;

    public boolean appliesTo(String json) {
        if (type != Type.CUSTOM) {
            throw new InvalidRouteException("Expression can be evaluated only for custom routes");
        }
        if (expression.isEmpty()) {
            throw new InvalidRouteException("Custom route must have an expression");
        }

        return expression.get().evaluate(json);
    }

    public boolean hasExpression() {
        return expression != null && expression.isPresent();
    }

    public static Route connect() {
        return connect(Collections.emptySet());
    }

    public static Route connect(Set<Backend<? extends BackendSettings>> backends) {
        return Route.builder()
                .type(Type.CONNECT)
                .name("Connect")
                .backends(backends)
                .expression(Optional.empty())
                .build();
    }

    public static Route disconnect() {
        return disconnect(Collections.emptySet());
    }

    public static Route disconnect(Set<Backend<? extends BackendSettings>> backends) {
        return Route.builder()
                .type(Type.DISCONNECT)
                .name("Disconnect")
                .backends(backends)
                .expression(Optional.empty())
                .build();
    }

    public static Route defaultRoute() {
        return defaultRoute(Collections.emptySet());
    }

    public static Route defaultRoute(Set<Backend<? extends BackendSettings>> backends) {
        return Route.builder()
                .type(Type.DEFAULT)
                .name("Default")
                .backends(backends)
                .expression(Optional.empty())
                .build();
    }

    public static Route custom(Expression<Boolean> expression) {
        return custom("Custom", Collections.emptySet(), expression);
    }

    public static Route custom(String name, Expression<Boolean> expression) {
        return custom(name, Collections.emptySet(), expression);
    }

    public static Route custom(
            String name, Set<Backend<? extends BackendSettings>> backends, Expression<Boolean> expression
    ) {
        return Route.builder()
                .type(Type.CUSTOM)
                .name(name)
                .backends(backends)
                .expression(Optional.ofNullable(expression))
                .build();
    }
}
